### PR TITLE
Nginx :  ngx_configure_listening_sockets and hijack `recvmsg`

### LINKED ISSUE #157: udp stream error
Compile nginx with --with-stream, and nginx.conf like this:
stream {
   upstream dns {
        server 192.168.1.3:53;
   }
  server {
       listen 53 udp;
       proxy_responses 1;
       proxy_timeout 20s;
       proxy_pass dns;
  }
}

This is the error message(error.log):
[alert] 16834#0 setsockopt(IP_PKTINFO) for 0.0.0.0:53 failed, ignore(9: Bad file descriptor)

### PR DESCRIPTION
**Pull request by firesidefox**
1. Nginx based on fstack delays setting up server on fstack until ngx_worker_process_init. ngx_configure_listening_sockets should as well.
2. FStack does not support IP_PKTINFO

Fix #157 .

### CODE REVIEW

File: app/nginx-1.11.10/src/core/ngx_connection.c [hunk 1]
```diff
@@ -13,6 +13,33 @@
...
+        if (!fstack_territory(ls[i].sockaddr->sa_family, ls[i].type, 0)) {    \
+            continue;                                                         \
+        }                                                                     \
+    } else {                                                                  \
+        ngx_log_error(NGX_LOG_ALERT, cycle->log, 0,                           \
+                            "unexpected process type: %d, ignored",           \
+                            ngx_process);                                     \
+        exit(1);                                                              \
+    }                                                                   
 #endif
```
Comment (whl739) [inline app/nginx-1.11.10/src/core/ngx_connection.c]: Is it better to use a function than macro here?

File: app/nginx-1.11.10/src/core/ngx_connection.c [hunk 2]
```diff
@@ -411,31 +438,7 @@ ngx_open_listening_sockets(ngx_cycle_t *cycle)
...
-                if (!fstack_territory(ls[i].sockaddr->sa_family, ls[i].type, 0)) {
-                    continue;
-                }
-            } else {
-                ngx_log_error(NGX_LOG_ALERT, cycle->log, 0,
-                                  "unexpected process type: %d, ignored",
-                                  ngx_process);
-                exit(1);
-            }
+            ngx_ff_skip_listening_socket(ls);
```
Comment (whl739) [inline app/nginx-1.11.10/src/core/ngx_connection.c]: It's better to pass `ls[i]` to `ngx_ff_skip_listening_socket`.

File: app/nginx-1.11.10/src/core/ngx_connection.c (hunk 1 above)
Comment (firesidefox) [reply app/nginx-1.11.10/src/core/ngx_connection.c]: On this, macro and function are same, encapsulates the public operation to carry out code reuse. 
Macro maybe have a better performance compared with (inline) function.  
What do you think? Have I missed something?

Comment (whl739) [reply app/nginx-1.11.10/src/core/ngx_connection.c]: > On this, macro and function are same, encapsulates the public operation to carry out code reuse.
Macro maybe have a better performance compared with (inline) function.

You are right, they are same, and macro has a better performance, but it has no effect, it's only called during initializing.
I just thought that the code is a little more here.

File: lib/ff_syscall_wrapper.c [hunk 3]
```diff
@@ -610,6 +610,11 @@ ff_sendmsg(int s, const struct msghdr *msg, int flags)
 {
     int rc;
 
+    if (msg->msg_name != NULL) {
+        linux2freebsd_sockaddr(msg->msg_name,
```
Comment (whl739) [inline lib/ff_syscall_wrapper.c]: `msg`'s value has been modified, it's against with the const qualifier.
Better like this:
```
void *oldsa = msg->msg_name;
struct sockaddr sa;
linux2freebsd_sockaddr(msg->msg_name, sizeof(struct linux_sockaddr), &sa);
msg->msg_name = &sa;
sendit..
msg->msg_name = oldsa;
```
or
```
struct msghdr m;
memcpy(&m, msg, sizeof(struct msghdr));
if (m.msg_name != NULL) {
    linux2freebsd_sockaddr(m->msg_name, sizeof(struct linux_sockaddr), m->msg_name);
}
```

**Pull request merged**